cpp/src/gandiva/llvm_ir_store.h

```diff
@@ -0,0 +1,49 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <arrow/status.h>
+#include <gandiva/visibility.h>
+#include <llvm/Support/MemoryBuffer.h>
+#include <memory>
+#include <vector>
+
+namespace gandiva {
+using arrow::Status;
+
+class GANDIVA_EXPORT LLVMIRStore {
+ public:
+  LLVMIRStore(const LLVMIRStore&) = delete;
+  LLVMIRStore& operator=(LLVMIRStore const&) = delete;
+
+  static LLVMIRStore* Get();
+
+  /// \brief add an LLVM IR to the store from a given bitcode file path
+  Status Add(const std::string& bitcode_file_path);
+
+  /// \brief add an LLVM memory buffer containing bitcode to the store
+  Status Add(std::unique_ptr<llvm::MemoryBuffer> buffer);
+
+  /// \brief get a list of LLVM memory buffers saved in the store
+  std::vector<std::unique_ptr<llvm::MemoryBuffer>>& GetIRBuffers();
```

Comment:
For safety and convenience, can you make the `Get` function private and `Add`/`GetIRBuffers` function static. So user would only need to call `LLVMIRStore::Add` and the global instance is hidden from user.
Also, is it possible to make `GetIRBuffers` return a const reference rather than a mutable reference?

Reply:
> can you make the Get function private and Add/GetIRBuffers function static?

Instead of using the singleton pattern, now I simply remove the `Get` API and make all public APIs in this class static, and access a static internal variables in llvm_external_ir_store.cpp file, as similar as function registry's way to store this state, could you please take a look and see if it fits? Thanks.

> is it possible to make GetIRBuffers return a const reference rather than a mutable reference?

Done

Reply:
Yeah LGTM. Thanks!